stan/math/opencl/kernel_generator/reduction_2d.hpp

```diff
@@ -0,0 +1,342 @@
+
+/**
+ * Represents a two dimensional reduction in kernel generator expressions. So as
+ * to be efficient two dimensional reductions are only done partially. That
+ * means instead of 1 element kernel output can have a few rows and a few
+ * columns that need to be reduced to obtain final result. This can be done in a
+ * separate kernel or after copying to CPU. Also two dimensional reductions can
+ * not be used as arguments to other operations - they can only be evaluated.
+ * @tparam Derived derived type
+ * @tparam T type of first argument
+ * @tparam Operation type with member function generate that accepts two
+ * variable names and returns OpenCL source code for reduction operation_cl
+ */
+template <typename Derived, typename T, typename Operation>
+class reduction_2d
+    : public internal::reduction_2d_base,
+      public operation_cl<Derived, typename std::remove_reference_t<T>::Scalar,
+                          T> {
+ public:
+  using Scalar = typename std::remove_reference_t<T>::Scalar;
+  using base = operation_cl<Derived, Scalar, T>;
+  using base::var_name_;
+  static const bool require_specific_local_size = true;
+
+ protected:
+  std::string init_;
+  using base::derived;
+
+ public:
+  using base::cols;
+  /**
+   * Constructor
+   * @param a the expression to reduce
+   * @param init OpenCL source code of initialization value for reduction
+   */
+  explicit reduction_2d(T&& a, const std::string& init)
+      : base(std::forward<T>(a)), init_(init) {}
+
+  /**
+   * Generates kernel code for assigning this expression into result expression.
+   * @param[in,out] generated map from (pointer to) already generated local
+   * operations to variable names
+   * @param[in,out] generated_all map from (pointer to) already generated all
+   * operations to variable names
+   * @param ng name generator for this kernel
+   * @param row_index_name row index variable name
+   * @param col_index_name column index variable name
+   * @param result expression into which result is to be assigned
+   * @return part of kernel with code for this and nested expressions
+   */
+  template <typename T_result>
+  kernel_parts get_whole_kernel_parts(
+      std::map<const void*, const char*>& generated,
+      std::map<const void*, const char*>& generated_all, name_generator& ng,
+      const std::string& row_index_name, const std::string& col_index_name,
+      const T_result& result) const {
+    kernel_parts parts = derived().get_kernel_parts(
+        generated, generated_all, ng, row_index_name, col_index_name, false);
+    kernel_parts out_parts = result.get_kernel_parts_lhs(
+        generated, generated_all, ng, row_index_name, col_index_name);
+
+    parts.args += out_parts.args;
+    parts.reduction_2d += "if (lid_i == 0) {\n"
+                     + result.var_name_
+                     + "_global[wg_id_j * n_groups_i + wg_id_i] = "
+                     + derived().var_name_ + "_local[0];\n"
+                     "}\n";
```

Comment:
Whenever I need to do this for debugging I add a print into `multi_result_kernel.hpp` at the end of `get_kernel_source_impl()`.

Reply:
Also I included a test that compares generated code with reference kernel, so you can also look at that when reviewing.

Reply:
Doesn't need to happen in this PR but it might be nice to have a `STAN_DEBUG_OPENCL_CODEGEN` macro or something like that name which just turns that on and prints out the kernel code

Reply:
I like the idea.